L81-core-cpp-jwt-token-audience.md

```diff
@@ -0,0 +1,128 @@
+/// (JWT) created with this credentials. It should not exceed
+/// \a kMaxAuthTokenLifetimeSecs or will be cropped to this value.
+/// user_provided_audience is an optional field for user to override the
+/// auidence in the JWT token. If user_provided_audience is empty, the service
+/// URL will be used as the audience.
+std::shared_ptr<CallCredentials> ServiceAccountJWTAccessCredentials(
+    const grpc::string& json_key,
+    long token_lifetime_seconds = kMaxAuthTokenLifetimeSecs,
+    const grpc::string& user_provided_audience = "");
+
+
+/// Builds credentials with reasonable defaults.
+/// user_provided_audience is an optional field for user to override the
+/// auidence in the JWT token. If user_provided_audience is empty, the service
+/// URL will be used as the audience.
+std::shared_ptr<ChannelCredentials> GoogleDefaultCredentials(
+    const grpc::string& user_provided_audience = "");
+```
+
+We also propose to change the C core API as follows
+
+```
+/** Creates a JWT credentials object. May return NULL if the input is invalid.
+   - json_key is the JSON key string containing the client's private key.
+   - token_lifetime is the lifetime of each Json Web Token (JWT) created with
+     this credentials.  It should not exceed grpc_max_auth_token_lifetime or
+     will be cropped to this value.
+   - user_provided_audience is an optional field for user to override the
+     auidence in the JWT token. If user_provided_audience is empty, the
+     service URL will be used as the audience.  */
+GRPCAPI grpc_call_credentials*
+grpc_service_account_jwt_access_credentials_create(
+    const char* json_key, gpr_timespec token_lifetime,
+    const char* user_provided_audience);
+
+/** Creates default credentials to connect to a google gRPC service.
+   WARNING: Do NOT use this credentials to connect to a non-google service as
+   this could result in an oauth2 token leak. The security level of the
+   resulting connection is GRPC_PRIVACY_AND_INTEGRITY.
+
+   If specified, the supplied call credentials object will be attached to the
+   returned channel credentials object. The call_credentials object must remain
+   valid throughout the lifetime of the returned grpc_channel_credentials
+   object. It is expected that the call credentials object was generated
+   according to the Application Default Credentials mechanism and asserts the
+   identity of the default service account of the machine. Supplying any other
+   sort of call credential will result in undefined behavior, up to and
+   including the sudden and unexpected failure of RPCs.
+
+   If nullptr is supplied, the returned channel credentials object will use a
+   call credentials object based on the Application Default Credentials
+   mechanism.
+
+   user_provided_audience is an optional field for user to override the
+   audience in the JWT token if used. If user_provided_audience is empty,
+   the service URL will be used as the audience.
+*/
```

Comment:
nit: similar to above, we should probably note in this API comment that user_provided_audience will only be used if a service account JWT access credential is created by the application default credentials mechanism

Also, let's note that `user_provided_audience` will be ignored if `call_credentials` is not nullptr (since then we will not use application default credentials, if I'm correct)

Reply:
Good suggestion. Comments updated.

If call_credentials it not nullptr, the input call_credentials will be used. Thus user_provided_audience will be ingored.